File: 2015/ruby/6/main-part-2.rb
#!/usr/bin/env ruby

def get_action(line)
    words = line.split(" ")
    if words[0] == "turn"
        action = words[1]
        start = words[2].split(",").map { |x| x.to_i }
        end_ = words[4].split(",").map { |x| x.to_i }
        return action, start, end_
    else
        action = words[0]
        start = words[1].split(",").map { |x| x.to_i }
        end_ = words[3].split(",").map { |x| x.to_i }
        return action, start, end_
    end
end

def main
    args = ARGV

    if args.length < 1
        puts "Usage: ruby main-part1.rb <input.txt>"
        exit 1
    end

    filename = args[0]
    file = File.open(filename, "r")

    if file.nil?
        puts "Could not open file"
        exit 1
    end

    lines = file.readlines
    file.close

    grid = Array.new(1000) { Array.new(1000, 0) }
    
    lines.each do |line|
        action, start, end_ = get_action(line)
        for i in start[0]..end_[0]
            for j in start[1]..end_[1]
                if action == "on"
                    grid[i][j] += 1
                elsif action == "off"
                    if grid[i][j] > 0
                        grid[i][j] -= 1
                    end
                else
                    grid[i][j] += 2
                end
            end
        end
    end

    count = 0
    for i in 0..999
        for j in 0..999
            count += grid[i][j]
        end
    end

    puts "Part 2: #{count}"
end

main()

File: 2015/python/6/main-part-2.py
#!/usr/bin/env python3

import sys
import os

def get_action(line):
    words = line.split()
    if words[0] == "turn":
        return {
            "action": words[1],
            "start": words[2].split(","),
            "end": words[4].split(",")
        }
    else:
        return {
            "action": words[0],
            "start": words[1].split(","),
            "end": words[3].split(",")
        }

def main():
    # get command line arguments
    if len(sys.argv) != 2:
        print("Usage: python3 main-part1.py <input.txt>")
        sys.exit(1)

    # get input file
    filename = sys.argv[1]

    # check if file exists
    if not os.path.exists(filename):
        print(f"Error: File '{filename}' not found")
        sys.exit(1)

    # read file
    lines = []
    with open(filename, 'r') as f:
        lines = f.readlines()

    grid = [[0 for i in range(1000)] for j in range(1000)]
    # get total area
    for line in lines:
        action = get_action(line)
        for i in range(int(action["start"][0]), int(action["end"][0]) + 1):
            for j in range(int(action["start"][1]), int(action["end"][1]) + 1):
                if action["action"] == "on":
                    grid[i][j] += 1
                elif action["action"] == "off":
                    grid[i][j] -= 1
                    if grid[i][j] < 0:
                        grid[i][j] = 0
                else:
                    grid[i][j] += 2

    # count total lights on
    total = 0
    for i in range(1000):
        for j in range(1000):
            total += grid[i][j]

    print("Part 2:", total)

if __name__ == "__main__":
    main()